components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,515 @@
+      return absl::nullopt;
+    }
+  } else {
+    n = input.length();
+  }
+
+  return std::make_pair(n, input.substr(field_end, n - field_end));
+}
+
+absl::optional<std::pair<std::string::size_type, std::string_view>>
+ExtractOptionalValue(std::string_view input,
+                     std::string_view field,
+                     bool is_resources = false) {
+  auto effective_field = base::StrCat({"\n", field});
+  auto field_end = effective_field.length();
+  if (field_end == input.length()) {
+    return absl::nullopt;
+  }
+  auto field_pos = input.rfind(effective_field);
+  if (field_pos == std::string::npos) {
+    return std::make_pair(std::string::npos, input);
+  }
+  if (field_pos != 0) {
+    return absl::nullopt;
+  }
+  std::string::size_type n;
+  if (is_resources) {
+    n = input.length();
+  } else {
+    // Check if this is the last optional field
+    n = input.find('\n', field_end);
+    if (n == std::string::npos) {
+      return std::make_pair(input.length(), input.substr(field_end));
+    }
+    if (n == field_end) {
+      return absl::nullopt;
+    }
+  }
+
+  return std::make_pair(n, input.substr(field_end, n - field_end));
+}
+
+absl::optional<std::string_view> GetNextSegment(const std::string& message,
+                                                size_t index) {
+  if (index > message.size() - 1) {
+    return absl::nullopt;
+  }
+  return std::string_view(message.begin() + index, message.end());
+}
+
+}  // namespace
+
+SIWEMessageParser::SIWEMessageParser() = default;
+SIWEMessageParser::~SIWEMessageParser() = default;
+
+mojom::SIWEMessagePtr SIWEMessageParser::Parse(const std::string& message) {
+  state_ = State::kStart;
+  auto resultt = mojom::SIWEMessage::New();
+  size_t index = 0;
+  if (!ParseSchemeAndDomain(message, index, resultt->origin)) {
+    return {};
+  }
+  auto next_segment = GetNextSegment(message, index);
+  if (!next_segment || !ParseAddress(*next_segment, index, resultt->address)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment ||
+      !ParseStatement(*next_segment, index, resultt->statement)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment || !ParseURI(*next_segment, index, resultt->uri)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment || !ParseVersion(*next_segment, index, resultt->version)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment || !ParseChainId(*next_segment, index, resultt->chain_id)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment || !ParseNonce(*next_segment, index, resultt->nonce)) {
+    return {};
+  }
+  next_segment = GetNextSegment(message, index);
+  if (!next_segment ||
+      !ParseIssuedAt(*next_segment, index, resultt->issued_at)) {
+    return {};
+  }
+  if (state_ == State::kOptionalFields) {
```

Comment:
So `state_` is read only once, it seems we don't need this field and its type at all now.